File: src/main/java/com/jigumulmi/common/PagedResponseDto.java
package com.jigumulmi.common;

import java.util.List;
import lombok.Builder;
import lombok.Getter;
import lombok.experimental.SuperBuilder;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

@Getter
@SuperBuilder
public class PagedResponseDto<T> {

    @Getter
    @Builder
    public static class PageDto {

        private Long totalCount;
        private int currentPage;
        private int totalPage;

        public static PageDto of(Page<?> page, Pageable pageable) {
            return PageDto.builder()
                .totalCount(page.getTotalElements())
                .currentPage(pageable.getPageNumber() + 1)
                .totalPage(page.getTotalPages())
                .build();
        }
    }

    private PageDto page;
    private List<T> data;

    public static <T> PagedResponseDto<T> of(Page<T> pageData, Pageable pageable) {
        return PagedResponseDto.<T>builder()
            .page(PageDto.of(pageData, pageable))
            .data(pageData.getContent())
            .build();
    }
}

File: admin/src/main/kotlin/com/jigumulmi/common/PagedResponseDto.kt
package com.jigumulmi.common

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable

open class PagedResponseDto<T>(
    open val page: PageDto,
    open val data: List<T>
) {
    data class PageDto(
        val totalCount: Long,
        val currentPage: Int,
        val totalPage: Int
    ) {
        companion object {
            fun of(page: Page<*>, pageable: Pageable): PageDto {
                return PageDto(
                    totalCount = page.totalElements,
                    currentPage = pageable.pageNumber + 1,
                    totalPage = page.totalPages
                )
            }
        }
    }

    companion object {
        fun <T> of(pageData: Page<T>, pageable: Pageable): PagedResponseDto<T> {
            return PagedResponseDto(
                page = PageDto.of(pageData, pageable),
                data = pageData.content
            )
        }
    }
}

